### ISSUE TITLE
tfcompile with tf.nn.dynamic_rnn crashes

### ISSUE THREAD

**Issue by carlthome** [tensorflow/tensorflow#11275]
Trying to build a C++ binary with tfcompile crashes with `INVALID ARGUMENTS: Mising Exit successor to rnn/while/Switch` if the graph contains `tf.nn.dynamic_rnn`, but it works with `tf.nn.static_rnn`. Why is this?

**Comment by drpngx** [comment]
@tatatodd any idea?

**Comment by tatatodd** [comment]
@carlthome What version of tensorflow are you using?  Since XLA and tfcompile support is currently experimental, I'll suggest you build from sources at HEAD.

Also, can you provide a minimal test case that demonstrates the problem?  Thanks!

**Comment by patrikohlsson** [comment]
I'm having the same issue with tfcompiler and tf.nn.dynamic_rnn.

OS:
Ubuntu 16.04
TF version:
1.2.1 built from source without GPU support and no XLA JIT.

Test case:
```python
xin = tf.placeholder(tf.float32, [1, 10, 10], name='input')
cell = tf.contrib.rnn.LSTMCell(512)
out, _ = tf.nn.dynamic_rnn(cell, xin, dtype=tf.float32)
out = tf.identity(out, name='output')

sess = tf.Session()
sess.run(tf.global_variables_initializer())

saver = tf.train.Saver()
saver.save(sess, 'outgraph')
```

outgraph.config.pbtxt:
```
feed {
  id {
    node_name: "input"
  }
  shape {
    dim { size: 1 }
    dim { size: 10 }
    dim { size: 10 }
  }
}

fetch {
  id {
    node_name: "output"
  }
}
```

BUILD file:
```python
load("//tensorflow/compiler/aot:tfcompile.bzl", "tf_library")

tf_library(
  name = "test_dyn_rnn",
  cpp_class = "TestDynRNN",
  graph = "outgraph.pb",
  config = "outgraph.config.pbtxt",
  freeze_checkpoint = "outgraph.ckpt"
)
```

Command:
`bazel build //tensorflow/compiler/aot/tests/test_dyn_rnn:test_dyn_rnn`

The build command results in `INVALID ARGUMENTS: Missing Exit successor to encoder/rnn/while/Switch`.
Swapping to `tf.nn.static_rnn` works however.

[Here's](https://gist.github.com/patrikohlsson/955a059d1f89c5339dd6d28d24c7f914) a complete example.

**Comment by carlthome** [comment]
I've been able to consistently reproduce the error `INVALID ARGUMENTS: Missing Exit successor to rnn/while/Switch` with master, thanks to @patrikohlsson's example. Perhaps @ebrevdo has some insight into what makes `tf.nn.dynamic_rnn` finicky with tfcompile?

**Comment by patrikohlsson** [comment]
With the XLA JIT fix posted by kayzhu in #11122 I can now reproduce the `Missing Exit successor` with XLA JIT enabled as well.

**Comment by tatatodd** [comment]
The `tf.nn.dynamic_rnn` library uses TensorFlow control-flow, which isn't well supported by XLA.  There's a little bit of special support in the JIT codepath, but the AOT (i.e. tfcompile) codepath doesn't support this at all right now.

Errors about `Enter`, `Exit`, `Switch` and `Merge` nodes are all symptoms of of this problem.

There is no workaround at the moment.  There are some in-process discussions and prototyping about how to handle this.  Our exact approach hasn't been fleshed out yet.

@hawkinsp may have some more details.

**Comment by carlthome** [comment]
Cool. It would be nice if XLA supported loops eventually. Is there a list in the documentation about what ops are not well supported at the moment? Particularly, the `tf.nn` stuff I'd expect to work as it's a core package.

For my particular use case with a recurrent neural network (RNN), it's fine to unroll it before tfcompile and pad/truncate test data as needed, but I'd imagine there are applications where that fix would not be straightforward.

**Comment by tatatodd** [comment]
@carlthome Yeah, one way or another we'll have support for loops in XLA AOT; we just need to work out the technical details.

We don't have documentation about what TensorFlow ops we support via XLA AOT or JIT at the moment, but the implementation of all of the TF->XLA translations is here:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/tf2xla/kernels

And yes, thanks for mentioning unrolling.  Indeed for some models, a workaround is to statically unroll your loops before feeding into XLA.  An example for LSTMs is here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tests/lstm.py
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tests/lstm_test.py

**Comment by ebrevdo** [comment]
Keep in mind that tf.nn.dynamic_rnn relies on tf.while_loop; there are now
some XLA primitives you can use (as of the last couple of days) in github
master; see, i.e.,

https://github.com/tensorflow/tensorflow/commit/3e9cd2e13da566e279396d232004d3c4ffad336e

https://github.com/tensorflow/tensorflow/commit/eb1fe50da445d3880b588215f6fadcc7f48dd3ff

You'll want to download a tensorflow nightly if you want to try it.

Keep in mind this code will perform a *full unroll* internally on the XLA
side, and nested while_loops (while_loop inside a dynamic_rnn) are not
really supported.

You want to be really careful with this new, relatively untested, xla
bridge.  See the new xla rnn unit tests for examples.

**Comment by jart** [comment]
@tatatodd I'm assigning to you to take out of triage queue. Feel free to re-assign

**Comment by carlthome** [comment]
ETA on tf.nn.dynamic_rnn support in XLA AOT?

**Comment by hawkinsp** [comment]
@carlthome For inference (not training), it might actually work right now. No-one has been working directly on AOT support for tf.nn.dynamic_rnn(), but as it happens I think all the pieces are there. Give it a go and let me know what happens!

**Comment by ebrevdo** [comment]
Don't we need the new max iterations attribute on while_loop?

**Comment by hawkinsp** [comment]
@ebrevdo I suspect not, actually. We need an iteration bound only for training (for Stacks as it happens, not for TensorArrays.)

**Comment by carlthome** [comment]
Well, new error at least so that's good:

`INVALID ARGUMENTS: Input 0 of node encoder/rnn/while/add/y was passed int32 from encoder/rnn/while/Switch:1 incompatible with expected INVALID.`

**Comment by hawkinsp** [comment]
@carlthome Yes, we're making progress :-) Would you be able to give me a small self-contained reproduction to look at? Thanks!

**Comment by carlthome** [comment]
Yeah! :man_dancing: 

Anyway from latest master with all defaults (e.g.
`git clone git@github.com:tensorflow/tensorflow.git && cd tensorflow && PYTHON_BIN_PATH=$(which python) USE_DEFAULT_PYTHON_LIB_PATH=1 CC_OPT_FLAGS='-march=native' TF_ENABLE_XLA=0 TF_NEED_MPI=0 TF_NEED_JEMALLOC=1 TF_NEED_GCP=0 TF_NEED_HDFS=0 TF_NEED_VERBS=0 TF_NEED_OPENCL=0 TF_NEED_CUDA=0 TF_NEED_GDR=0 ./configure`)

And this is basically just @patrikohlsson's example:
## create_graph.py:
```py
import tensorflow as tf

x = tf.placeholder(tf.float32, [1, 10, 10], name='input')
c = tf.nn.rnn_cell.LSTMCell(32)
y, _ = tf.nn.dynamic_rnn(c, x, dtype=tf.float32)
y = tf.identity(y, name='output')

with tf.Session() as session:
  session.run(tf.global_variables_initializer())
  saver = tf.train.Saver(write_version=tf.train.SaverDef.V1)
  saver.save(session, 'graph.ckpt')
```
## graph.config.pbtxt:
```
feed {
  id { node_name: "input" }
  shape {
    dim { size: 1 }
    dim { size: 10 }
    dim { size: 10 }
  }
}
fetch {
  id { node_name: "output" }
}
```
## BUILD:
```
load("//tensorflow/compiler/aot:tfcompile.bzl", "tf_library")

tf_library(
  name = "dynamic_rnn_graph",
  cpp_class = "Graph",
  graph = "graph.pb",
  config = "graph.config.pbtxt",
  freeze_checkpoint = "graph.ckpt"
)
```

Assuming TensorFlow is installed, run `create_graph.py` and put its output files into the repo root together with `BUILD` and `graph.config.pbtxt`, and then run `bazel build :dynamic_rnn_graph`.

**Comment by lissyx** [comment]
Looking at the codebase, r1.4 includes this https://github.com/tensorflow/tensorflow/commit/0e71ecaf9512cd8a69af01ac85e5e1632171c651 which is mostly what we already had as a workaround on https://github.com/mozilla/tensorflow/ repo.

**Comment by carlthome** [comment]
I tried again just now with master and still failing with `INVALID ARGUMENTS: Input 0 of node encoder/rnn/while/add/y was passed int32 from encoder/rnn/while/Switch:1 incompatible with expected INVALID.`

**Comment by carlthome** [comment]
Ping @ebrevdo, could we expect `tf.nn.dynamic_rnn` to work with XLA eventually?

**Comment by ebrevdo** [comment]
Yes; though I don't promise it'll happen by end of December.

**Comment by alquraishi** [comment]
@ebrevdo do you expect to see any speed gains?

**Comment by leod** [comment]
As a follow-up question, is it within XLA's goals to work with seq2seq models at some point? Will it be possible to use XLA on graphs that have variable length inputs and outputs? Right now, it seems that `tfcompile` does not like fetches with dimension `-1`.

**Comment by ebrevdo** [comment]
It is a goal; but because XLA will JIT a graph and perform memory layout
*before* any execution, you will have to provide a "maximum max_time" for
the memory layout (even if the loop will not calculate up to this maximum
size).  When working with dynamic_rnn, for example, this will mean you will
have to pad your batched sequence inputs to some specific max_time size(s),
like 10, 50, 100, ...; to ensure XLA doesn't try to JIT compile for every
possible max(sequence_length) across your minibatches.  The tensorflow NMT
tutorial shows how to bucket batches by sequence lengths, but does not
currently pad the input max_times to the bucket boundaries.

**Comment by leod** [comment]
Sounds good, the requirement to pad sequences makes sense. Thank you for the quick reply!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by lissyx** [comment]
@ebrevdo What happens for a `tf.nn.bidirectional_dynamic_rnn` with no `sequence_length` argument? I am currently removing it because `tfcompile` chokes on `Assert`. This used to work quite well, but since r1.4, compiling with the recurrent layers, even very small ones, fails in a way that looks like infite loop consuming as much as CPU and memory as possible. Last time I tried master, it was exposing the same behavior.

My guess is that this is because of the lack of `sequence_length`, and thus it goes weird.

However, keeping the `sequence_length` fails this way:
```
2018-01-05 18:29:44.901953: F external/org_tensorflow/tensorflow/compiler/aot/tfcompile_main.cc:140] Non-OK-status: status status: Not found: Op type not registered 'Assert' in binary running on hostname. Make sure the Op and Kernel are registered in the binary running in this process.
```

**Comment by ebrevdo** [comment]
+todd; who I think can say more about tfcompile's interactions with
tf.Assert.

**Comment by tatatodd** [comment]
Last I heard, @prb12 was working on adding a no-op tf2xla kernel for tf.Assert, he would know more.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lissyx** [comment]
@ebrevdo So, it looks like with current master, it's not an issue anymore ? At least I can again use `tfcompile`

**Comment by ebrevdo** [comment]
Yes; we've fixed this up recently.  I'll close this bug so if you run into
any issues please open a new one.

**Comment by carlthome** [comment]
Nice! Looking forward to testing this again. 🎉

**Comment by lissyx** [comment]
@ebrevdo Is it supposed to be fixed on r1.5 as well ?

**Comment by ebrevdo** [comment]
I think not, only nightlies and the upcoming 1.6
